edb/lib/std/70-constructorfuncs.eql

```diff
@@ -0,0 +1,527 @@
+        ELSE
+            to_number(s, fmt)
+        END
+    )
+    $$;
+};
+
+
+CREATE FUNCTION
+std::int64(s: std::str, fmt: OPTIONAL str={}) -> std::int64
+{
+    FROM SQL $$
+    SELECT (
+        CASE WHEN fmt IS NULL THEN
+            s::bigint
+        ELSE
+            to_number(s, fmt)::bigint
+        END
+    )
+    $$;
+};
+
+
+CREATE FUNCTION
+std::int32(s: std::str, fmt: OPTIONAL str={}) -> std::int32
+{
+    FROM SQL $$
+    SELECT (
+        CASE WHEN fmt IS NULL THEN
+            s::int
+        ELSE
+            to_number(s, fmt)::int
+        END
+    )
+    $$;
+};
+
+
+CREATE FUNCTION
+std::int16(s: std::str, fmt: OPTIONAL str={}) -> std::int16
+{
+    FROM SQL $$
+    SELECT (
+        CASE WHEN fmt IS NULL THEN
+            s::smallint
+        ELSE
+            to_number(s, fmt)::smallint
+        END
+    )
+    $$;
+};
+
+
+CREATE FUNCTION
+std::float64(s: std::str, fmt: OPTIONAL str={}) -> std::float64
+{
+    FROM SQL $$
+    SELECT (
+        CASE WHEN fmt IS NULL THEN
+            s::float8
+        ELSE
+            to_number(s, fmt)::float8
+        END
+    )
+    $$;
+};
+
+
+CREATE FUNCTION
+std::float32(s: std::str, fmt: OPTIONAL str={}) -> std::float32
+{
+    FROM SQL $$
+    SELECT (
+        CASE WHEN fmt IS NULL THEN
+            s::float4
+        ELSE
+            to_number(s, fmt)::float4
+        END
+    )
+    $$;
+};
+
+
+# std::bool
+# --------
+CREATE FUNCTION
+std::bool(val: std::bool) -> std::bool
+{
+    FROM SQL $$
+    SELECT val
+    $$;
+};
+
+
+CREATE FUNCTION
+std::bool(val: std::str) -> std::bool
+{
+    FROM SQL $$
+    SELECT val != ''
+    $$;
+};
+
+
+CREATE FUNCTION
+std::bool(val: std::bytes) -> std::bool
+{
+    FROM SQL $$
+    SELECT val != ''::bytea
+    $$;
+};
+
+
+CREATE FUNCTION
+std::bool(val: std::int64) -> std::bool
+{
+    FROM SQL $$
+    SELECT val != 0
+    $$;
+};
+
+
+CREATE FUNCTION
+std::bool(val: std::float64) -> std::bool
+{
+    FROM SQL $$
+    SELECT val != 0
+    $$;
+};
+
+
+CREATE FUNCTION
+std::bool(val: std::decimal) -> std::bool
+{
+    FROM SQL $$
+    SELECT val != 0
+    $$;
+};
+
+
+CREATE FUNCTION
+std::bool(val: array<anytype>) -> std::bool
+{
+    FROM SQL $$
+    SELECT COALESCE(array_length(val, 1), 0) != 0
+    $$;
+};
+
+
+CREATE FUNCTION
+std::bool(val: std::timedelta) -> std::bool
+{
+    FROM SQL $$
+    SELECT date_part('epoch', val) != 0
+    $$;
+};
+
+
+# Without a bool taking time explicitly, Postgres cannot choose
+# between bool(interval) and bool(any).
+CREATE FUNCTION
+std::bool(val: std::naive_time) -> std::bool
+{
+    FROM SQL $$
+    SELECT TRUE
+    $$;
+};
+
+
+CREATE FUNCTION
+std::bool(val: std::json) -> std::bool
+{
+    FROM SQL $$
+    SELECT val NOT IN (
+        'null'::jsonb, '""'::jsonb, '[]'::jsonb, '{}'::jsonb, 'false'::jsonb,
+        '0'::jsonb
+    )
+    $$;
+};
+
+
+# Any scalar that's not explicitly covered is "true" by virtue of existing.
+CREATE FUNCTION
+std::bool(val: anyscalar) -> std::bool
```

Comment:
I don't think a catch-all converter is a good idea.  If there is no explicit conversion, it is undefined.

Reply:
> that's why we're introducing the bool constructor to get the truthy values of many more things.

Not really. The point of the constructors is to allow parametrization, not necessarily to extend the domain of inputs.

Reply:
> Not really. The point of the constructors is to allow parametrization, not necessarily to extend the domain of inputs.

Contructors in general, sure. However, it appears that `bool` in particular is mainly serving the purpose of allowing a larger range of types to be treated as familiar truthy values similar to other languages. Unfortunately, there's no consistent agreement on truthiness among other languages (say Python, JS, and C) even though there are some common practices (such as treating numerical 0 as equivalent to "false" and non-zero as "true").

I think it's convenient to extend `bool` to every type because then you could ask "is a particular arbitrary set truthy". This can come up as a potentially useful `FILTER`.

```
SELECT Foo
FILTER Foo.bar != '' AND Foo.baz > 0;

# vs

SELECT Foo
FILTER bool(Foo.bar) AND bool(Foo.baz);

# or even

SELECT Foo
FILTER all({bool(Foo.bar), bool(Foo.baz)});
```
Imagine more than 2 links/props that are used in the same fashion in the above example and the general application of `bool` conveys the intent more clearly than explicit listing of various possibilities, I think.

Reply:
After a discussion with @elprans and @1st1 we came to the conclusion that `bool` as truthiness is not as good a feature as it first appeared.